src/Microsoft.Azure.WebJobs.Host/Executors/FunctionExecutor.cs

```diff
@@ -73,7 +73,59 @@ public HostOutputMessage HostOutputMessage
         public async Task<IDelayedException> TryExecuteAsync(IFunctionInstance functionInstance, CancellationToken cancellationToken)
         {
             var functionInstanceEx = (IFunctionInstanceEx)functionInstance;
-            var logger = _loggerFactory?.CreateLogger(LogCategories.CreateFunctionCategory(functionInstance.FunctionDescriptor.LogName));
+            var logger = _loggerFactory.CreateLogger(LogCategories.CreateFunctionCategory(functionInstanceEx.FunctionDescriptor.LogName));
+
+            if (functionInstance.FunctionDescriptor.RetryStrategy != null)
+            {
+                return await ExecuteWithRetriesAsync(functionInstanceEx, cancellationToken, logger, functionInstance.FunctionDescriptor.RetryStrategy);
+            }
+
+            return await TryExecuteAsyncCore(functionInstanceEx, cancellationToken, logger);
+        }
+
+        internal async Task<IDelayedException> ExecuteWithRetriesAsync(IFunctionInstanceEx functionInstance, CancellationToken token, ILogger logger, IRetryStrategy retryStrategy)
+        {
+            var attempt = 0;
+            IDelayedException functionResult = null;
+            TimeSpan nextDelay;
+            do
+            {
+                if (token.IsCancellationRequested)
+                {
+                    logger.LogDebug($"Function execution cancelled.");
+                    break;
+                }
+                logger.LogFunctionRetryAttempt(attempt, retryStrategy.MaxRetryCount);
```

Comment:
Let's change this log so that it's only emitted if there is actually a failure, so in normal non-failure cases, we emit the same number of logs that we did before (for performance reasons).